_posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md

```diff
@@ -0,0 +1,132 @@
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+Wait a second, what have we done here?
+
+We define the first version of the `maybe_attack` function to take in a first and second argument named `last_shooter`. This means the function will only match if the player provided as a first argument matches the player provided as a second argument
+Thanks to immutability, when we call both arguments by the same name Elixir will check if they are actually the same!
+We could easily call both arguments player or even something like player_is_the_last_shooter. It doesn't matter! The rule is just that if we want to ensure that there is equality, we call both arguments by the same name!
+
+**Ok, it's time to play using our nice little code!**
+
+Let's say we have `player1` and `player2`, `player1` made the most recent attack and therefore is our last_shooter and now `player2` will try to attack!
+
+So we will call the function `maybe_attack(player2, player1)` where `player2` is the player who wants to make an attack and `player1` is our last_shooter
+
+We have two `maybe_attack` functions both with arity 2, so Elixir will try to pattern match from top to bottom, i.e. the first function it will try to match will be 
+
+```elixir
+def maybe_attack(last_shooter, last_shooter) do
+    ignore()
+end
+```
+
+Our first argument is `player2` and Elixir will bind it with `last_shooter = player2` and since our second argument is also a `last_shooter`, Elixir will use the `^` (pin operator) to check if the previous bind is valid for the second argument instead of trying to rebind it.
+
+```elixir
+last_shooter = player2
+ˆlast_shooter = player1
+```
+
+As player2 is different from player1, we will not have a valid pattern match and therefore Elixir will move on to try to match with the next function!
+
+**Our next match attempt!**
+
+```elixir
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+Now the behavior will be different, we are asking Elixir to match two different arguments. That is, to make a bind for each one.
+
+```elixir
+player = player2
+last_shooter = player1
+```
+
+With a valid match, our function will run! Player2 will make an attack and then will be registered as our new `last_shooter`!
+
+What if player2 tries another attack in a row?
+Well, we will call our first `maybe_attack` function again and try a match. Player2 wants to make an attack and is also `last_shooter`, so we get the following call
+
+`maybe_attack(player2, player2)`
+
+Trying to match it with the first maybe_attack function we get the following match
+
+```elixir
+def maybe_attack(last_shooter, last_shooter) do
+    ignore()
+end
+```
+
+```elixir
+last_shooter = player2
+ˆlast_shooter = player2
+```
+
+Which is a valid match! And since our function just ignores the attack attempt, nothing will happen until another player attempts an attack!
+
```

Comment:
What do you think about including a couple sentences here to summarize and wrap up the blog post? Something like:

> That's it! We've learned how pattern matching and data immutability together can provide an elegant solution to control flows, another tool in our Elixir toolbox!